libs/@local/hash-graph-sdk/typescript/src/entity.ts

```diff
@@ -1,14 +1,124 @@
 import type { Entity as GraphApiEntity } from "@local/hash-graph-client/api";
-import type { EntityId } from "@local/hash-graph-types/entity";
+import type {
+  EntityId,
+  EntityMetadata,
+  EntityPropertiesObject,
+  LinkData,
+} from "@local/hash-graph-types/entity";
 
-export class Entity {
-  #entity: GraphApiEntity;
+const typeId: unique symbol = Symbol.for(
+  "@local/hash-graph-sdk/entity/SerializedEntity",
+);
+type TypeId = typeof typeId;
 
-  constructor(entity: GraphApiEntity) {
-    this.#entity = entity;
+export interface SerializedEntity<
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  Properties extends EntityPropertiesObject = EntityPropertiesObject,
+> {
+  // Prevents the type from being created from outside the module
+  [typeId]: TypeId;
+}
+
+type EntityData<Properties extends EntityPropertiesObject> = {
+  metadata: EntityMetadata;
+  properties: Properties;
+  linkData?: LinkData;
+};
+
+type EntityInput<Properties extends EntityPropertiesObject> =
+  | GraphApiEntity
+  | SerializedEntity<Properties>;
+
+const isSerializedEntity = <Properties extends EntityPropertiesObject>(
+  entity: EntityInput<Properties>,
+): entity is SerializedEntity => {
+  return (
+    "entityTypeId" in
+    (entity as GraphApiEntity | EntityData<Properties>).metadata
+  );
+};
+
+const isGraphApiEntity = <Properties extends EntityPropertiesObject>(
+  entity: EntityInput<Properties>,
+): entity is GraphApiEntity => {
+  return (
+    "entityTypeIds" in
+    (entity as GraphApiEntity | EntityData<Properties>).metadata
+  );
+};
+
+export class Entity<
+  Properties extends EntityPropertiesObject = EntityPropertiesObject,
+> {
+  #entity: EntityData<Properties>;
```

Comment:
Is there any reason to not make this `private`?

Reply:
Yes, `#entity` also has runtime privacy while `private entity` is only for type checks. See https://github.com/hashintel/hash/pull/4503#discussion_r1612964485